deep/trackers.py

```diff
@@ -0,0 +1,127 @@
+from typing import Type
+from enum import Enum, auto, unique
+
+from django.core.cache import cache
+from django.utils import timezone
+from django.db import transaction, models
+from celery import shared_task
+from dateutil.relativedelta import relativedelta
+
+from deep.caches import CacheKey
+
+from user.models import Profile
+from project.models import Project
+
+
+class TrackerAction:
+    @unique
+    class Project(Enum):
+        READ = auto()
+        WRITE = auto()
+
+
+def track_project(project: Project, action: TrackerAction.Project = TrackerAction.Project.READ):
+    cache_key = CacheKey.Tracker.LAST_PROJECT_READ_ACCESS_DATETIME
+    if action == TrackerAction.Project.WRITE:
+        cache_key = CacheKey.Tracker.LAST_PROJECT_WRITE_ACCESS_DATETIME
+    cache.set(
+        cache_key + str(project.pk),
+        timezone.now(),
+        None,  # Cache value forever until manually removed
+    )
+
+
+def track_user(user_profile: Profile):
+    cache.set(
+        CacheKey.Tracker.LAST_USER_ACTIVE_DATETIME + str(user_profile.pk),
+        timezone.now(),
+        None,  # Cache value forever until manually removed
+    )
+
+
+@transaction.atomic
+def update_entity_data_in_bulk(
+    Model: Type[models.Model],
+    cache_key_prefix: str,
+    field: str,
+):
+    cache_keys = cache.keys(cache_key_prefix + '*')
+    entities_update = []
+    for key, value in cache.get_many(cache_keys).items():
+        entities_update.append(
+            Model(**{
+                'id': key.split(cache_key_prefix)[1],
+                field: value,
+            })
+        )
+    if entities_update:
+        Model.objects.bulk_update(entities_update, fields=[field], batch_size=200)
+    transaction.on_commit(
+        lambda: cache.delete_many(cache_keys)
+    )
+
+
+def update_project_data_in_bulk():
+    # -- Read
+    update_entity_data_in_bulk(
+        Project,
+        CacheKey.Tracker.LAST_PROJECT_READ_ACCESS_DATETIME,
+        'last_read_access',
+    )
+    # -- Write
+    update_entity_data_in_bulk(
+        Project,
+        CacheKey.Tracker.LAST_PROJECT_WRITE_ACCESS_DATETIME,
+        'last_write_access',
+    )
+    # -- Update project->status using last_write_access
+    # -- -- To active
+    Project.objects.filter(
+        last_write_access__lte=timezone.now() - relativedelta(years=1),
```

Comment:
Maybe we can have a constant like `PROJECT_INACTIVE_AFTER_YEARS` or something like that.

Reply:
Done https://github.com/the-deep/server/commit/3a19f3fd39d0d5d09253f94d32acdb1c4db0f953